sagemaker-python-sdk/tensorflow_script_mode_training_and_serving/mnist.py

```diff
@@ -0,0 +1,184 @@
+      train_op = optimizer.minimize(
+          loss=loss,
+          global_step=tf.train.get_global_step())
+      return tf.estimator.EstimatorSpec(mode=mode, loss=loss, train_op=train_op)
+
+    # Add evaluation metrics (for EVAL mode)
+    eval_metric_ops = {
+        "accuracy": tf.metrics.accuracy(
+            labels=labels, predictions=predictions["classes"])}
+    return tf.estimator.EstimatorSpec(
+        mode=mode, loss=loss, eval_metric_ops=eval_metric_ops)
+
+def _load_training_data(base_dir):
+    x_train = np.load(os.path.join(base_dir, 'train_data.npy'))
+    y_train = np.load(os.path.join(base_dir, 'train_labels.npy'))
+    return x_train, y_train
+
+def _load_testing_data(base_dir):
+    x_test = np.load(os.path.join(base_dir, 'eval_data.npy'))
+    y_test = np.load(os.path.join(base_dir, 'eval_labels.npy'))
+    return x_test, y_test
+
+def _parse_args():
+
+    parser = argparse.ArgumentParser()
+
+    # Data, model, and output directories
+    # model_dir is always passed in from SageMaker. By default this is a S3 path under the default bucket.
+    parser.add_argument('--model_dir', type=str)
+    parser.add_argument('--sm-model-dir', type=str, default=os.environ.get('SM_MODEL_DIR'))
+    parser.add_argument('--train', type=str, default=os.environ.get('SM_CHANNEL_TRAINING'))
+    parser.add_argument('--hosts', type=list, default=json.loads(os.environ.get('SM_HOSTS')))
+    parser.add_argument('--current-host', type=str, default=os.environ.get('SM_CURRENT_HOST'))
+
+    return parser.parse_known_args()
+
+def serving_input_fn():
+    inputs = {'x': tf.placeholder(tf.float32, [None, 784])}
+    return tf.estimator.export.ServingInputReceiver(inputs, inputs)
+
+if __name__ == "__main__":
+    args, unknown = _parse_args()
+
+    train_data, train_labels = _load_training_data(args.train)
+    eval_data, eval_labels = _load_testing_data(args.train)
+
+    # Create the Estimator
+    mnist_classifier = tf.estimator.Estimator(
+        model_fn=cnn_model_fn, model_dir=args.model_dir)
+
+    # Set up logging for predictions
+    # Log the values in the "Softmax" tensor with label "probabilities"
+    tensors_to_log = {"probabilities": "softmax_tensor"}
+    logging_hook = tf.train.LoggingTensorHook(
+        tensors=tensors_to_log, every_n_iter=50)
+
+    # Train the model
+    train_input_fn = tf.estimator.inputs.numpy_input_fn(
+        x={"x": train_data},
+        y=train_labels,
+        batch_size=100,
+        num_epochs=None,
+        shuffle=True)
+
+    # Evaluate the model and print results
+    eval_input_fn = tf.estimator.inputs.numpy_input_fn(
+        x={"x": eval_data},
+        y=eval_labels,
+        num_epochs=1,
+        shuffle=False)
+
+    train_spec = tf.estimator.TrainSpec(train_input_fn, max_steps=20000)
+    eval_spec = tf.estimator.EvalSpec(eval_input_fn)
+    tf.estimator.train_and_evaluate(mnist_classifier, train_spec, eval_spec)
+
+    if args.current_host == args.hosts[0]:
```

Comment:
I this line necessary? I thought that TF would handle it without that.

Reply:
Yeah. this is necessary. If we configure an exporter export model during evaluation TF handles but calling export_savedmodel TF won't check if the host is master or not.